File: leetcode_java/16. 3Sum Closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        long res = (long)Integer.MAX_VALUE;
        for (int k = 0; k < nums.length - 2; k++) {
            if(k > 0 && nums[k] == nums[k - 1]) continue;
            int i = k + 1, j = nums.length - 1;
            while(i < j) {
                int tmp = nums[k] + nums[i] + nums[j];
                if(Math.abs(tmp - target) < Math.abs(res - target)) res = tmp;
                if(tmp > target) while(i < j && nums[j] == nums[--j]);
                else if(tmp < target) while(i < j && nums[i] == nums[++i]);
                else return target;
            }
        }
        return (int)res;
    }
}

File: leetcode_python/16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: [int], target: int) -> int:
        nums.sort()
        res = float("inf")
        for k in range(len(nums) - 2):
            if k > 0 and nums[k] == nums[k - 1]: continue
            i, j = k + 1, len(nums) - 1
            while i < j:
                tmp = nums[k] + nums[i] + nums[j]
                if abs(tmp - target) < abs(res - target): res = tmp
                if tmp > target: j -= 1
                elif tmp < target: i += 1
                else: return target
        return res